File: Java/백준/Gold/2342. Dance Dance Revolution/Dance Dance Revolution.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    static int[][][] dp;
    static int[] arr;
    static int len = 0;
    static final int INF = 1000000000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        arr = new int[100001];
        StringTokenizer st = new StringTokenizer(br.readLine());
        int idx = 0;
        while (true) {
            int n = Integer.parseInt(st.nextToken());
            if (n == 0) break;
            arr[idx++] = n;
        }
        len = idx;
        dp = new int[len + 1][5][5];

        for (int i = 0; i <= len; i++) {
            for (int j = 0; j < 5; j++) {
                for (int k = 0; k < 5; k++) {
                    dp[i][j][k] = INF;
                }
            }
        }

        dp[0][0][0] = 0;

        for (int i = 0; i < len; i++) {
            int next = arr[i];
            for (int l = 0; l < 5; l++) {
                for (int r = 0; r < 5; r++) {
                    int now = dp[i][l][r];
                    if (now == INF) continue;
                    dp[i + 1][next][r] = Math.min(dp[i + 1][next][r], now + getCost(l, next));
                    dp[i + 1][l][next] = Math.min(dp[i + 1][l][next], now + getCost(r, next));
                }
            }
        }

        int answer = INF;
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                answer = Math.min(answer, dp[len][i][j]);
            }
        }

        System.out.println(answer);
    }

    static int getCost(int from, int to) {
        if (from == 0) return 2;
        if (from == to) return 1;
        if (Math.abs(from - to) == 2) return 4;
        return 3;
    }
}

File: C11/백준/Gold/2342. Dance Dance Revolution/Dance Dance Revolution.c
#include <stdio.h>
#include <stdlib.h>
#define INF 1000000000

int dp[100001][5][5];
int arr[100001];

int cost(int from, int to) {
    if (from == 0) return 2;
    if (from == to) return 1;
    if (abs(from - to) == 2) return 4;
    return 3;
}

int min(int a, int b) {
    return a < b ? a : b;
}

int main() {
    int n, idx = 0;
    while (1) {
        scanf("%d", &n);
        if (n == 0) break;
        arr[idx++] = n;
    }

    for (int i = 0; i <= idx; i++) {
        for (int j = 0; j < 5; j++) {
            for (int k = 0; k < 5; k++) {
                dp[i][j][k] = INF;
            }
        }
    }

    dp[0][0][0] = 0;

    for (int i = 0; i < idx; i++) {
        int next = arr[i];
        for (int l = 0; l < 5; l++) {
            for (int r = 0; r < 5; r++) {
                if (dp[i][l][r] == INF) continue;
                int moveLeft = dp[i][l][r] + cost(l, next);
                int moveRight = dp[i][l][r] + cost(r, next);
                if (moveLeft < dp[i + 1][next][r])
                    dp[i + 1][next][r] = moveLeft;
                if (moveRight < dp[i + 1][l][next])
                    dp[i + 1][l][next] = moveRight;
            }
        }
    }

    int answer = INF;
    for (int l = 0; l < 5; l++) {
        for (int r = 0; r < 5; r++) {
            if (dp[idx][l][r] < answer)
                answer = dp[idx][l][r];
        }
    }

    printf("%d\n", answer);
    return 0;
}